### ISSUE TITLE
Horizontal selection gets cut off

### ISSUE THREAD

**Issue by gaearon** [bvaughn/react-devtools-experimental#9]
Not sure when it happens but eventually the horizontal selection doesn't apply to the whole line:

![screen recording 2019-02-27 at 04 43 pm](https://user-images.githubusercontent.com/810438/53507537-ff64e180-3aaf-11e9-8f3c-19845e21ac98.gif)

**Comment by bvaughn** [comment]
I think this is being caused by the `box-sizing: border-box` style, but I'm not sure why.

**Comment by bvaughn** [comment]
This is a minimal repro:
https://codesandbox.io/s/8143655xvl

**Comment by rickbeerendonk** [comment]
Remove `width: 100%` (basically `width: initial`) will solve your sandbox.

**Comment by BolajiOlajide** [comment]
or setting `width: auto`.

**Comment by john-raymon** [comment]
I may be wrong but I don't see anything wrong with it being cut off ? Isn't it suppose to behave the way it is, overflowing it's text it's content over if `box-sizing` is set to something like `border-box` with padding such as the one in the repo ? Why don't you change the default `border-box` being applied and explicitly use `box-sizing: content-box` to https://codesandbox.io/s/8143655xvl

**Comment by stevenla** [comment]
`width: 100%` is [defined to be 100% of the container's width](https://developer.mozilla.org/en-US/docs/Web/CSS/width#Values). In your sandbox (and in the devtools), the width of the container is the visible viewport, so the row takes that width. The text hanging off the side is overflow, and not contributing to the size of the row.

<img width="733" alt="screen shot 2019-02-27 at 3 41 18 pm" src="https://user-images.githubusercontent.com/887038/53530865-2b7b6500-3aa6-11e9-8f21-6b74cf4a9179.png">

Making `width: initial` would work, but you'd lose out on the rounded corners on your selection (like on the left)

<img width="276" alt="screen shot 2019-02-27 at 3 42 59 pm" src="https://user-images.githubusercontent.com/887038/53530924-7301f100-3aa6-11e9-92d7-036e8cdd1d8e.png">

**Comment by bvaughn** [comment]
Unfortunately, removing width (or setting min width, or width "auto" or "initial") won't fully fix the problem, although my initial Code Sandbox wasn't complete (so I've updated it).

**Comment by fatfisz** [comment]
That might be going in a different direction, but I here's my take on this: https://codesandbox.io/s/2336y3xj30

Basically the OuterBox is immobilized and InnerBox has `overflow: auto;` instead (that could also work by wrapping OuterBox in another box). Then the Highlight thing is positioned absolutely wrt the most-outer box. I figured this will be simple to implement since `top` positions of Rows are known.

**Comment by bvaughn** [comment]
Hey! Cool idea @fatfisz! Lemme play with this some in the context of the larger project to figure out how easy it would be manually position the highlight row... I think it could work though!

**Comment by bvaughn** [comment]
Rats. Unfortunately, I think that having the inner box scroll horizontally and the outer box scroll vertically (which is necessary for the way our windowing component works) won't quite work because of awkward scrollbar placement (unless I'm doing something silly).

![sss](https://user-images.githubusercontent.com/29597/53538672-8ec7c000-3ac3-11e9-94c0-2291bf291c82.gif)

https://codesandbox.io/s/7m832z986

**Comment by rickyruiz** [comment]
> Hey folks! Thanks for the comments and ideas!
> 
> Unfortunately, removing width (or setting min width, or width "auto" or "initial") won't fully fix the problem, although my initial Code Sandbox wasn't complete (so I've updated it).
> 
> I might end up needing to use a JS solution for this (just keep track of the widest thing I've measured so far and set the inner width to that) but I was really hoping to avoid it because of the added complexity.

I did not understand why you cannot remove `width` or use `min-width` in your demo. Could you elaborate or provide a more complex demo?

___

In the meantime, here's an experiment with a drawback, it requires a fixed-width font (monospace) ... it works with minimal JS complexity, a CSS pseudo-element, a CSS variable, CSS calc function, and the CSS `ch` unit.

Looks for the row that has the max `text.length + paddingLeft` and uses it to set the css variable width value used in the pseudo-element (which acts as a highlight).

https://codesandbox.io/s/k2mo08q173

**Comment by solkimicreb** [comment]
My solution: https://codesandbox.io/s/6jzjmmyvm3 (Let me know if I missed something, I feel like others came up with this too.)

A few things I noticed:

- The issue is not really with `box-sizing` but with the `width` usage instead of `min-width` inside the rows. I think you would like to have your rows to be **at least the width** of the container and potentially longer if the content doesn't fit without scrolling. Simply replacing `width: 100%` with `min-width: 100%` will do this for you.

- You should leave the `box-sizing` rule as it is. It may seem like the issue is with `box-sizing` because the default `content-box` value only counts the content inside the width and leaves the padding out. This means that without the rule the `width: 100%` rows will have a `100% + {padding}px` width. This is exactly why nobody uses the default `box-sizing: content-box` rule. It is really hard to work with.

- Combining `box-sizing: border-box` with `min-width: 100%` will do what you are looking for in a predictable way.

- I don't see the need for both an outer and an inner box but it is probably something you need in the real scenario and not in the minimal repro. Anyways I would either remove the inner box (if it is not needed) or add a `position: relative;` rule to it as well to prevent future issues with the absolute positioned rows.

I hope these helped.

**Comment by solkimicreb** [comment]
I think I misunderstood above, sorry 🙁 If you would like to have all rows to fill the full width of the scroll container (even when scrolled) I think you will hit some heavy walls. This is a solution which does that: https://codesandbox.io/s/k5z957p7wo

It requires you to move away from absolute positioned rows though. The issue, in this case, is that you fill the inner box with only absolutely positioned rows which are outside of the layout flow (because of their absolute positioning.) This means that the inner box will never pick up the dimensions from its content.

You have two options from here. You either look out for the largest absolute positioned child inside the box with js and set the box width to the width of this child, or you can **move away from absolute positioning** so the box container can pick up the dimensions of its children. My example does the latter.

Sadly I don't think you will find a clever trick that solves the issue without one of the above for you.

**Comment by InvictusMB** [comment]
If there are no objections against flexbox this is my solution:
`.OuterBox` has to be `display: flex` so that `.InnerBox` grows to content size.
`.Row` has to remain in the flow and spread the `.InnerBox`.

`position: relative` lets `.Row` remain in the flow: https://codesandbox.io/s/0q9o0531rn
Or with if spacing is used instead of positioning: https://codesandbox.io/s/5w7kry2onp

**Comment by fatfisz** [comment]
I think the requirement is to have rows with `position: absolute` because of how react-window works.

**Comment by bvaughn** [comment]
Hey folks. Thanks for the additional comments. I'm in meetings for a while but I will look through the new sandbox links shortly!

Unfortunately the solutions that require moving away from absolute positioning or using flex box aren't options for us, given that the list is being rendered by [react-window](https://react-window.now.sh).

> @bvaughn Why can't you remove `width` or use `min-width` in your demo? Could you elaborate or provide a more complex demo?

Because then the background won't expand far enough for the narrower width elements.

**Comment by im-ironclad** [comment]
I've left a tweet, I believe that solves the problem unless I am not understanding it correctly. https://twitter.com/im_ironclad/status/1101191688272445441

**Comment by bvaughn** [comment]
> I've left a tweet, I believe that solves the problem unless I am not understanding it incorrectly. https://twitter.com/im_ironclad/status/1101191688272445441

Yeah, switching to min-width is an improvement. Still leaves something to be desired though, as [mentioned above](https://github.com/bvaughn/react-devtools-experimental/issues/9#issuecomment-468081954), since the narrower elements won't stretch to fill the full space

![screen shot 2019-02-28 at 10 54 37 am](https://user-images.githubusercontent.com/29597/53590724-494cd680-3b47-11e9-8513-5b204570b619.png)

**Comment by im-ironclad** [comment]
> > I've left a tweet, I believe that solves the problem unless I am not understanding it incorrectly. https://twitter.com/im_ironclad/status/1101191688272445441
> 
> Yeah, switching to min-width is an improvement. Still leaves something to be desired though, as [mentioned above](https://github.com/bvaughn/react-devtools-experimental/issues/9#issuecomment-468081954), since the narrower elements won't stretch to fill the full space
> 
> ![screen shot 2019-02-28 at 10 54 37 am](https://user-images.githubusercontent.com/29597/53590724-494cd680-3b47-11e9-8513-5b204570b619.png)

I would like to say with your current markup if you want the "shorter" elements to stretch the whole way you would have to do some math and explicitly set a width. That width would come from the width of your longest element assuming you can access it. A temporary solution so you can kind of see what I mean is to add min-width: 200% to the .row which will make the short ones 200% length no matter what. That may be incorrect, i'll think on it some more and check back in if there still isn't a full solution. (About to be busy for the afternoon)

**Comment by bvaughn** [comment]
>I would like to say with your current markup if you want the "shorter" elements to stretch the whole way you would have to do some math and explicitly set a width, or min-width actually.

Yeah, that's kind of [the conclusion I came to earlier as well](https://github.com/bvaughn/react-devtools-experimental/issues/9#issuecomment-468081954) unfortunately. ☹️

**Comment by im-ironclad** [comment]
> > I would like to say with your current markup if you want the "shorter" elements to stretch the whole way you would have to do some math and explicitly set a width, or min-width actually.
> 
> Yeah, that's kind of [the conclusion I came to earlier as well](https://github.com/bvaughn/react-devtools-experimental/issues/9#issuecomment-468081954) unfortunately. ☹️

I feel like i've solved something similar to this several times without JS, it's possible. Are you using SCSS? I'll get back to you.

**Comment by bvaughn** [comment]
> Are you using SCSS?

Nope. Combination of [CSS](https://github.com/bvaughn/react-devtools-experimental/blob/master/src/devtools/views/Element.css) and inline styles (abs positioning) from `react-window`.

**Comment by bvaughn** [comment]
Switched to a `min-width` approach in c7f8557. This isn't perfect, but it's an improvement at least.

**Comment by fatfisz** [comment]
If you have nothing against it I'll try to do it by using my approach and submit a PR. I'm sure there's a non-complicated way to make it work, since it's similar to what's being used in various editors already.

**Comment by bvaughn** [comment]
At this point, I'm mostly looking for bug repros (as GH issues) because I have a lot of features left to implement and not a lot of bandwidth to review PRs. If you're able to find a fix that doesn't require a lot of complex changes, please feel free to open a PR though.

**Comment by bvaughn** [comment]
[@GabbeV_](https://twitter.com/GabbeV_) pointed out a possible approach on Twitter that [uses relative positioning and negative margins](https://twitter.com/GabbeV_/status/1101250469853302784) to stretch the background highlight.

Unfortunately this looks to cause undesirable behavior when combined with `scrollIntoView`:
![fnky](https://user-images.githubusercontent.com/29597/53603968-fcc5c300-3b67-11e9-99cf-644bb074852b.gif)

Here's a patch against 90001e8 showing what I tried:
```diff
diff --git a/src/devtools/views/Element.css b/src/devtools/views/Element.css
index 6227ee6..4dafb11 100644
--- a/src/devtools/views/Element.css
+++ b/src/devtools/views/Element.css
@@ -1,7 +1,6 @@
 .Element,
 .SelectedElement {
   border-radius: 0.25em;
-  position: relative;
   white-space: nowrap;
   line-height: var(--line-height-data);
   display: flex;
diff --git a/src/devtools/views/Element.js b/src/devtools/views/Element.js
index ca26ebf..16aa0d4 100644
--- a/src/devtools/views/Element.js
+++ b/src/devtools/views/Element.js
@@ -91,6 +91,8 @@ export default function ElementView({ index, style }: Props) {
         paddingLeft: `${(depth - baseDepth) * 0.75 + 0.25}rem`,
         width: undefined,
         minWidth: '100%',
+        position: 'relative',
+        marginBottom: `-20px`,
       }}
     >
       <span className={styles.Component} ref={ref}>
diff --git a/src/devtools/views/Tree.js b/src/devtools/views/Tree.js
index da32139..d9ffdf1 100644
--- a/src/devtools/views/Tree.js
+++ b/src/devtools/views/Tree.js
@@ -101,6 +101,7 @@ export default function Tree(props: Props) {
             <FixedSizeList
               className={styles.List}
               height={height}
+              innerElementType={innerElementType}
               itemCount={numElements}
               itemData={itemData}
               itemSize={lineHeight}
@@ -115,3 +116,7 @@ export default function Tree(props: Props) {
     </div>
   );
 }
+
+const innerElementType = ({ style, ...rest }) => (
+  <div style={{ ...style, display: 'inline-block', minWidth: '100%', width: undefined }} {...rest} />
+);
\ No newline at end of file
```

**Comment by GabbeV** [comment]
@bvaughn Seems like the rows in the sandbox were a different height.
Changing margin-bottom from "-20px" to -style.height solves the shaking.

**Comment by bvaughn** [comment]
😆 Damn, I feel foolish. I misremembered my own row height.

**Comment by bvaughn** [comment]
Looks like you're right @GabbeV – this seems to work! I feel a little weird about it but for now, I'll roll forward with it. 😄 Thanks!

**Comment by bvaughn** [comment]
Fixed in 8aa6f96
